runtime/j9vm/jvm.c

```diff
@@ -452,6 +456,74 @@ static const J9SignalMapping signalMap[] = {
 	{NULL, J9_SIG_ERR}
 };
 
+static void
+captureCommandLine(void)
+{
+#define ENV_VAR_NAME "IBM_JAVA_COMMAND_LINE"
+
+#if defined(LINUX)
+	int fd = open("/proc/self/cmdline", O_RDONLY, 0);
+
+	if (fd >= 0) {
+		char *buffer = NULL;
+		size_t length = 0;
+		for (;;) {
+			char small_buffer[512];
+			ssize_t count = read(fd, small_buffer, sizeof(small_buffer));
+			if (count <= 0) {
+				break;
+			}
+			length += (size_t)count;
+		}
+		if (length < 2) {
+			goto done;
+		}
+		/* final NUL is already included in length */
+		buffer = malloc(length);
+		if (NULL == buffer) {
+			goto done;
+		}
+		if ((off_t)-1 == lseek(fd, 0, SEEK_SET)) {
+			goto done;
+		}
+		if (read(fd, buffer, length) != length) {
+			goto done;
+		}
+		/* replace the internal NULs with spaces */
+		for (length -= 2;; length -= 1) {
+			if (0 == length) {
+				break;
+			}
+			if ('\0' == buffer[length]) {
+				buffer[length] = ' ';
+			}
+		}
+		/* it's not fatal if setenv() fails, so don't bother checking */
+		setenv(ENV_VAR_NAME, buffer, 0 /* don't overwrite */);
+		free(buffer);
+done:
+		close(fd);
+	}
+#elif defined(WIN32) /* defined(LINUX) */
+	size_t existingLength = 0;
+
+	if (0 != getenv_s(&existingLength, NULL, 0, ENV_VAR_NAME)) {
+		/* getenv_s() failed */
+	} else if (0 != existingLength) {
+		/* variable is already defined */
+	} else {
+		const char *commandLine = GetCommandLineA();
+
+		if (NULL != commandLine) {
+			/* it's not fatal if _putenv_s() fails, so don't bother checking */
+			_putenv_s(ENV_VAR_NAME, commandLine);
```

Comment:
Is there a wide version that can be used instead, along with GetCommandLineW()?

Reply:
I'm not sure anything else is prepared for wide characters, but I'll investigate.

Reply:
Using wide ensures we preserve all characters correctly in the environment variable. It doesn't affect fetching the environment variable. Any place that fetches the value can continue to work as it does now. Windows does any necessary conversion between wide and the format requested.

Reply:
Fixed in 229e47cda1681285905983db3863a8c897ba9eab.